app/domain/authentication/authn_azure/validate_application_identity.rb

```diff
@@ -9,119 +9,158 @@ module AuthnAzure
-          service_id:       @service_id,
-          logger:           @logger
-        )
+      # add prefix to all permitted constraints
+      def prefixed_permitted_constraints prefix
+        AZURE_RESOURCE_TYPES.map { |k| "#{prefix}#{k}" }
       end
 
       def validate_required_constraints_exist
-        validate_constraint_exists :subscription_id
-        validate_constraint_exists :resource_group
+        validate_resource_constraint_exists "subscription-id"
+        validate_resource_constraint_exists "resource-group"
       end
 
-      def validate_constraint_exists constraint
-        unless application_identity.constraints[constraint]
-          raise Err::RoleMissingConstraint, annotation_type_constraint(constraint)
-        end
+      def validate_resource_constraint_exists resource_type
+        resource = application_identity.resources.find { |a| a.resource_type == resource_type }
+        raise Err::RoleMissingConstraint, resource_type unless resource
       end
 
-      # validates that the application identity doesn't include logical constraint
+      # validates that the application identity doesn't include logical resource constraint
       # combinations (e.g user_assigned_identity & system_assigned_identity)
       def validate_constraint_combinations
-        identifiers = %i(user_assigned_identity system_assigned_identity)
+        identifiers = %w(user-assigned-identity system-assigned-identity)
 
-        identifiers_constraints = application_identity.constraints.keys & identifiers
+        identifiers_constraints = application_identity.resource_types & identifiers
         unless identifiers_constraints.length <= 1
-          raise Errors::Authentication::IllegalConstraintCombinations, annotation_type_constraints(identifiers_constraints)
+          raise Errors::Authentication::IllegalConstraintCombinations, identifiers_constraints
         end
       end
 
-      def validate_token_identity_matches_annotations
-        application_identity.constraints.each do |constraint|
-          annotation_type  = constraint[0].to_s
-          annotation_value = constraint[1]
-          unless annotation_value == @token_identity[annotation_type.to_sym]
-            raise Err::InvalidApplicationIdentity, annotation_type_constraint(annotation_type)
-          end
+      # xms_mirid is a term in Azure to define a claim that describes the resource that holds the encoding of the instance's
+      # among other details the subscription_id, resource group, and provider identity needed for authorization.
+      # xms_mirid is one of the fields in the JWT token. This function will extract the relevant information from
+      # xms_mirid claim and populate a representative hash with the appropriate fields.
+      def extract_resources_from_token
+        parse_xms_mirid
+        @resources_from_token = [
+          AzureResource.new(
+            resource_type: "subscription-id",
+            resource_name: xms_mirid.subscriptions
```

Comment:
what if xms_mirid.subscriptions doesn't exist?

Reply:
then we won't get to this line. We validate the required fields in the XmsMirid object creation.

Reply:
you right you right